gtsam/sfm/ShonanAveraging.cpp

```diff
@@ -819,20 +835,33 @@ template class ShonanAveraging<2>;
 
 ShonanAveraging2::ShonanAveraging2(const Measurements &measurements,
                                    const Parameters &parameters)
-    : ShonanAveraging<2>(measurements, parameters) {}
+    : ShonanAveraging<2>(parameters.useHuber
+                             ? makeNoiseModelRobust(measurements)
+                             : measurements,
+                         parameters) {}
+
 ShonanAveraging2::ShonanAveraging2(string g2oFile, const Parameters &parameters)
-    : ShonanAveraging<2>(parseMeasurements<Rot2>(g2oFile), parameters) {}
+    : ShonanAveraging<2>(
+          parameters.useHuber
+              ? makeNoiseModelRobust(parseMeasurements<Rot2>(g2oFile))
+              : parseMeasurements<Rot2>(g2oFile),
+          parameters) {}
 
 /* ************************************************************************* */
 // Explicit instantiation for d=3
 template class ShonanAveraging<3>;
 
 ShonanAveraging3::ShonanAveraging3(const Measurements &measurements,
                                    const Parameters &parameters)
-    : ShonanAveraging<3>(measurements, parameters) {}
+    : ShonanAveraging<3>(parameters.useHuber?
+        makeNoiseModelRobust(measurements) : measurements, parameters) {}
 
 ShonanAveraging3::ShonanAveraging3(string g2oFile, const Parameters &parameters)
-    : ShonanAveraging<3>(parseMeasurements<Rot3>(g2oFile), parameters) {}
+    : ShonanAveraging<3>(
+          parameters.useHuber
+              ? makeNoiseModelRobust(parseMeasurements<Rot3>(g2oFile))
```

Comment:
Create a parseMeasurements version that takes the flag rather than complicate the logic of these constructors

Reply:
`parseMeasurements` is a part of `dataset.cpp` whereas `makeNoiseModelRobust` is of `ShonanAveraging.h`. This kind of specialization of `parseMeasurements` seems finicky to me.

Reply:
Just add it here as a local helper method

Reply:
I came up with something neater: `maybeRobust(measurements, useHuber)`. If `useHuber` is false, just return the measurements.

Reply:
This also make the constructor more functional, since we can use it across all constructors as simple as `maybeRobust(measurements)` or `maybeRobust(parseMeasurements<Rot3>(g2oFile))`.